src/sentry/search/snuba/backend.py

```diff
@@ -193,8 +194,66 @@ def query(
         date_from=None,
         date_to=None,
     ):
-        from sentry.models import Group, GroupStatus, GroupSubscription
+        search_filters, paginator_options, retention_window_start = self.initialize_variables(
```

Comment:
I don't understand.

It could be overriden in the future.

When I did less refactoring, it seemed you were suggesting to do more so that I would be less likely to have to touch the base class again. Did I misunderstand?

Much of the work you're telling me to remove was towards that goal - passing all the params, in case in the future they are needed. Initializing as a function in case we need to add more in the future, etc...

Reply:
I think there is some misunderstanding here, let me elaborate those principles now that we have examples. I am not surprised we are not 100% in alignment considering all the ground we covered in a 30 minutes chat.

Principles around what we discussed yesterday:

**General Idea:** Designing and refactoring is striking a balance between competing and conflicting goals. Generally picking only one principle and going all in on that everywhere does not help. This is specifically true in na case like this where the class is lacking design and structure.

**Responsibility of the parent class** It is true I suggested doing refactoring to touch the parent class less, but that has to be put in context, we were talking about the query execution chunk of this class. This case is very different and other goals are more interesting:
- for query execution we already know for a fact that it has to work differently in the two implementation. So we are 100% certain that if we do not move the whole responsibility in the subclass we will have to continuously update the parent class when we make changes in the logic of one of the implementations. This case is the opposite, not only we do not have evidence that this logic needs to be overridden. It is actually very unlikely, so preserving clarity seems to have a higher value over extensibility. The logic is also simple enough that, in the rare case it will have to be changed in a subclass we will refactor it then.
- Deciding when to introduce an abstraction, moving code into a function adds and abstraction (doing it only for the sake of reusing a piece of code is generally a smell, because it can make code less readable). The abstraction is supposed to simplify the understanding of the class and to define the design. A good rule of thumb to understand whether, through a method, you are adding a useful abstraction is asking yourself what the role of that abstraction is in the domain model. This class performs a search. Good abstractions are "paginating results" "query the backend", "serialize output", "split query into smaller ones", etc. These actions have a well defined meaning in the domain model, they are somewhat independent from each other (they reader can reason about them in isolation), and they communicate clearly what they would do in the context of the system. "Initialize variables" is the opposite: it is extremely vague, it has no meaning in the context of the work the class is doing unless you look into the implementation. So this does not seem a useful abstraction to add. 

**The 'might need it later' approach** While future proofing is a very good idea in general, it can spectacularly backfire if it is used in the wrong place. "Might need it later" is only 30% of the question to ask when deciding whether to add something you do not need immediately. The two followup questions are:
- how likely will it be I will need it later
- what's the cost of introducing something now vs what is likely going to be the cost later.

Let's take the example of the query logic we discussed to move to the subclasses yesterday:
- might need it later? Yes 
- how likely will it be? We already need it now so 100%
- what's the cost of not introducing it: awful. We know that the alternative is adding a lot of additional code paths to the main query process that is already very complex.

Let's take the example of variable initialization
- might need it later? yes
- how likely it is it will be needed later? seems unliklely to me. The logic in the method is fairly generic and the return values you foresee are tightly coupled with the logic itself.
- what's the cost of adding it later? Seems very small. because these are jsut a handful of unrelated statements.

**Adding unused parameters** Adding an abstraction by introducing a function is adding a constraint to the scope that function has access to. This is something positive, not a limitation. Limiting the scope means defining it, thus it means making the code easier to understand, it defines responsibilities and it lets the design emerge. Passing to a function more than it needs is defying the purpose of introducing the abstraction since the function is not constrained anymore to operate on what you designed it for. Thus it takes away the advantages. So please remove them. 
When we discussed about passing to the query function all it needed so that it would not access the state of the parent class directly, that implied also restricting the list of parameters to what it needed, otherwise we are not limiting nor defining what the function needs.
Taking a step back on this, if you feel that a function you are introducing may be overridden in the future and may need more parameters than it actually does now, then this may mean that the responsibility of the function you are introducing is not well defined enough. It should be taken as a warning.

Happy to chat more in person to develop the subject anytime